spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/quartz/TracingJobListenerTest.java

```diff
@@ -0,0 +1,330 @@
+				.build();
+
+		listener = new TracingJobListener(tracing);
+		completableJob = new CompleteableTriggerListener();
+
+		scheduler = createScheduler(getClass().getSimpleName(), 1);
+		scheduler.addJob(newJob(ExceptionalJob.class).withIdentity(EXCEPTIONAL_JOB_KEY)
+				.storeDurably().build(), true);
+		scheduler.addJob(newJob(SuccessfulJob.class).withIdentity(SUCCESSFUL_JOB_KEY)
+				.storeDurably().build(), true);
+
+		scheduler.getListenerManager().addTriggerListener(listener);
+		scheduler.getListenerManager().addJobListener(listener);
+		scheduler.getListenerManager()
+				.addTriggerListener((CompleteableTriggerListener) completableJob);
+		scheduler.getListenerManager()
+				.addJobListener((CompleteableTriggerListener) completableJob);
+		scheduler.start();
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		scheduler.shutdown(true);
+	}
+
+	@Test
+	public void should_return_class_name_all_the_time() {
+		// when
+		String name = listener.getName();
+
+		// expect
+		assertThat(name).isEqualTo(TracingJobListener.class.getName());
+	}
+
+	@Test
+	public void should_complete_span_when_job_is_successful() throws Exception {
+		// given
+		Trigger trigger = newTrigger().forJob(SUCCESSFUL_JOB_KEY).startNow().build();
+
+		// when
+		runJob(trigger);
+
+		// expect
+		takeSpan();
+	}
+
+	@Test
+	public void should_have_span_with_proper_name_and_tag_when_job_is_successful()
+			throws Exception {
+		// given
+		Trigger trigger = newTrigger().withIdentity(TRIGGER_KEY)
+				.forJob(SUCCESSFUL_JOB_KEY).startNow().build();
+
+		// when
+		runJob(trigger);
+
+		// expect
+		Span span = takeSpan();
+		assertThat(span.name()).isEqualToIgnoringCase(SUCCESSFUL_JOB_KEY.toString());
+		assertThat(span.tags().get(TRIGGER_TAG_KEY))
+				.isEqualToIgnoringCase(TRIGGER_KEY.toString());
+	}
+
+	@Test
+	public void should_complete_span_when_job_throws_exception() throws Exception {
+		// given
+		Trigger trigger = newTrigger().forJob(EXCEPTIONAL_JOB_KEY).startNow().build();
+
+		// when
+		runJob(trigger);
+
+		// expect
+		takeSpan();
+	}
+
+	@Test
+	public void should_complete_span_when_job_is_vetoed() throws Exception {
+		// given
+		scheduler.getListenerManager().addTriggerListener(new VetoJobTriggerListener());
+		Trigger trigger = newTrigger().forJob(SUCCESSFUL_JOB_KEY).startNow().build();
+
+		// when
+		runJob(trigger);
+
+		// expect
+		takeSpan();
+	}
+
+	@Test
+	public void should_not_complete_span_when_context_is_modified_to_remove_keys()
+			throws Exception {
+		// given
+		scheduler.getListenerManager().addJobListener(new ContextModifyingJobListener());
+		Trigger trigger = newTrigger().forJob(EXCEPTIONAL_JOB_KEY).startNow().build();
+
+		// when
+		runJob(trigger);
+
+		// expect
+		requireNoSpan();
+	}
+
+	@Test
+	public void should_have_parent_and_child_span_when_trigger_contains_span_info()
+			throws Exception {
+		// given
+		brave.Span span = tracing.tracer().nextSpan();
+		JobDataMap data = new JobDataMap();
```

Comment:
suppose the real question is if JobDataMap is type bound consistently from the job. if that's the case, then we can't stick strings in it. need some real world stuff I guess. excellent discovery

Reply:
So I understand that doing the `instanceof String` is good enough for this usecase?

Reply:
Yea ... I've added a test case to cover the concern where the `JobDataMap` is constructed with a map that does not have a compatible typing. As expected, the code still works since `JobDataMap` internally is using a `Map<String, Object>` and all values from the map you give in the constructor to `JobDataMap` are simply copied over.